src/words.ts

```diff
@@ -0,0 +1,147 @@
+import {mmod} from 'xen-dev-utils';
+
+type StepVector = Record<string, number>;
+
+function sum(array: number[]) {
+  let sum = 0;
+  array.forEach(x => {
+    sum += x;
+  });
+  return sum;
+}
+
+export function zeroVector(): StepVector {
+  return {};
+}
+
+export function add(v1: StepVector, v2: StepVector): StepVector {
+  // Clone because we don't want to mutate the input.
+  const result = {...v1};
+  for (const [letter, value] of Object.entries(v2)) {
+    // For each entry,
+    // if `result` does not have the entry then add the component from v2 as an entry.
+    if (!(letter in v1)) {
+      result[letter] = value;
+    }
+    // If `result` has a nonzero coefficient for that entry:
+    else {
+      // If the sum at the component is 0, then delete the entry.
+      if (value + v1[letter] === 0) {
+        delete result[letter];
+      } else {
+        // Otherwise update the component to the sum.
+        result[letter] += value;
+      }
+    }
+  }
+  return result;
+}
+
+export function scalarMult(scalar: number, v: StepVector): StepVector {
+  if (scalar === 0) {
+    return zeroVector();
+  } else {
+    // Clone because we don't want to mutate the input.
+    const result = {...v};
+    for (const letter of Object.keys(result)) {
+      result[letter] *= scalar;
+    }
+    return result;
+  }
+}
+
+// Additive inverse of a `StepVector`
+export function neg(v: StepVector): StepVector {
+  return scalarMult(-1, v);
+}
+
+// Subtract two `StepVector`s
+export function sub(v1: StepVector, v2: StepVector): StepVector {
+  return add(v1, neg(v2));
+}
+
+// Rotate a string, moving it left by `offset`.
+export function rotate(str: string, offset: number): string {
+  return [...Array(str.length).keys()]
+    .map(i => str[mmod(i + offset, str.length)])
+    .join('');
+}
+
+// Elements in the resulting StepVector have the order given by `orderedLetterList`.
+export function getStepVector(word: string, intervalClass: number): StepVector {
+  if (word.length === 0) {
+    // Always return the zero vector if `word` is empty
+    return zeroVector();
+  } else if (intervalClass % word.length === 0) {
+    return scalarMult(
+      Math.floor(intervalClass / word.length),
+      stepSignature(word)
+    );
+  } else {
+    // Initialize `result` to a prefix with `intervalClass / word.length` letters
+    const result = scalarMult(
+      Math.floor(intervalClass / word.length),
+      stepSignature(word)
+    );
+    const slice = String(word).substring(0, intervalClass % word.length);
```

Comment:
JS has three ways of doing the same thing: `.slice`, `.substr` and `.substring`.
Prior art in this repository uses `.slice` so it would be preferrable here too for style consistency.